Product Collection: carousel layout frontend

kmanijak opened the pull request:
### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

<!-- You can erase any parts of this template not applicable to your Pull Request. -->

### Changes proposed in this Pull Request:

<!-- If necessary, indicate if this PR is part of a bigger feature. Add a label with the format `focus: name of the feature [team:name of the team]`. -->

<!-- Describe the changes made to this Pull Request and the reason for such changes. -->

<!-- For bug fixes: If known, please provide links to help with traceability and escape analysis. -->
<!-- Please include a link to the issue of the bug being fixed, if one doesn't exist please create it. -->
<!-- If the PR that introduced the bug is known, please also add its link below. -->

This PR covers the frontend experience of the Product Collection carousel layout. That includes:
- proper Next/Prev Buttons positioning
- adding scrolling logic,
- enabling/disabling buttons,
- hiding Next/Prev Buttons:
    - if there's no scroll (adjusted in Product Gallery as well if there's no images to scroll)
    - on mobile.

New Next/Prev position is at the top of Product Template (right side). If there's Heading, they're put together in Row block and aligned.

Closes https://github.com/woocommerce/woocommerce/issues/59404

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

<!-- If this PR includes UI changes, please provide screenshots or a screen recording for clarity. -->
<!-- This section can be removed if not applicable. -->

| Desktop | Mobile |
| ------ | ----- |
| https://github.com/user-attachments/assets/cc0913d8-1500-4ae3-89bb-54b827ede6f8 | https://github.com/user-attachments/assets/af8d7f54-f90b-4967-847d-92391d30d9c9 | 

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

#### Positioning, scrolling and disabling

1. Create new post
2. Add Product Collection block and choose "create your own" so there's lots of products to display
3. Change the layout to Carousel
4. **Expected:** Buttons are placed in top-right corner in Row block
5. Set "Products in Carousel" to some higher number, e.g. 20
6. Save and go to frontend
7. **Expected:**
    - Previous button id disabled by default
    - Position is preserved
8. Use buttons to scroll through carousel
9. **Expected:** Each click scrolls nearly whole width of carousel (90%)
10. Scroll to the very end
11. **Expected:** Next button gets disabled
12. Scroll back to the very beginning
13. **Expected:** Previous button gets disabled

#### Positioning 2

1. Create new post
2. Add Product Collection block and choose "On Sale Products" 
3. Change the layout to Carousel
4. **Expected:** Buttons are placed in top-right corner in Row block with Heading

#### Hide when not necessary

1. Create new post
2. Add Product Collection block and choose create your own so there's lots of products to display
3. Change the layout to Carousel
6. Set "Products in Carousel" to some small number, so products don't exceed the carousel (e.g. 2)
7. Add Product block and choose some product that has 0 or 1 images
8. Add Product Gallery inside
9. Save and go to frontend
10. **Expected:**
    - There's no arrows in Product Gallery
    - There's no arrows in Product Collection carousel *

> [!NOTE]
> * Arrows are initially _rendered and disappear_ in Product Collection. This is due to fact we're determining it based on rendered content and it takes a moment for iAPI to process initial context in frontend. I decided to put initial state to show arrows as I assume carousel will more often used to show _more_ products exceeding the screen width rather than less.

#### Hide on mobile

1. Repeat steps 1 - 7 from previous section but:
    - choose product in Product block with at least 2 images
    - set lots of products in Product Collection, e.g. 20
2. Set the mobile view on the frontend
3. **Expected:**
    - Arrows are not displayed in neither of blocks
4. Use gestures to scroll carousel
6. **Expected:** It scrolls through and snaps to product

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

kmanijak committed:
Move Product Gallery Next Prev block from inner blocks to main blocks directory and rename to NextPrevButtons

kmanijak committed:
Rename frontend block to NextPrevButtons

kmanijak committed:
Update blocks classnames

kmanijak committed:
Move styles from Product Gallery to standalone block

kmanijak committed:
Update block references to be Next Previous Buttons

kmanijak committed:
Remove obsolete E2E tests

kmanijak committed:
Add Product Gallery E2E test about opinionated mobile layout

kmanijak committed:
Update webpack entries

kmanijak committed:
Add chanbgelog

kmanijak committed:
Update assets paths

kmanijak committed:
Merge branch 'trunk' into wooplug-4986-extract-next-prev-buttons

kmanijak committed:
Make sure block is usable in Product Collection

kmanijak committed:
Make actions more abstract

kmanijak committed:
Provide external context

kmanijak committed:
Remove postId from usesContext

kmanijak committed:
Merge branch 'trunk' into wooplug-4986-extract-next-prev-buttons

kmanijak committed:
Make Next Previous Buttons accepting explicit iAPI provider

kmanijak committed:
Add basic actions in Product Collection

kmanijak committed:
Add initial state

kmanijak committed:
Add changelog

kmanijak committed:
Add simple scrolling logic to Product Collection for demo purposes

kmanijak committed:
Fix lint

kmanijak committed:
Fix another lint

kmanijak committed:
Merge branch 'trunk' into wooplug-4986-extract-next-prev-buttons

kmanijak committed:
Update tests

kmanijak committed:
Merge branch 'wooplug-4986-extract-next-prev-buttons' into wooplug-4986-generic-block-i

Aljullu committed:
Use CSS variables for Product Gallery sizes

Aljullu committed:
Add changelog file

kmanijak committed:
Merge branch 'update/product-gallery-css-variables' into wooplug-4986-generic-block-i

kmanijak committed:
Revert "Use CSS variables for Product Gallery sizes"

This reverts commit 080eb70eb4a7a172f9147b12336934553812059e.

kmanijak committed:
Revert "Add changelog file"

This reverts commit 6c3557fe078426c07d40b2fa95b4481f12ee9f2d.

kmanijak committed:
Merge branch 'trunk' into wooplug-4986-generic-block-i

kmanijak committed:
Handle Next/Prev Buttons positioning

kmanijak committed:
Hide scrollbar on carousel

kmanijak committed:
Scroll carousel by 90% width of container

kmanijak committed:
Hide arrows on mobile

kmanijak committed:
Provide opinionated width styles

kmanijak committed:
Handle arrows disabling

kmanijak committed:
Add scroll watch directive to disable buttons if needed

kmanijak committed:
Add logic to disable buttons when we're on either side of scroll

kmanijak committed:
Add logic to hide arrows if not needed

kmanijak committed:
Remove the logic of hiding arrows in Large Image

kmanijak committed:
Fix linter

kmanijak committed:
Merge branch 'trunk' into wooplug-4885-product-collection-carousel-layout-frontend-logic

kmanijak committed:
Fix lint

kmanijak committed:
Add changelog

kmanijak committed:
improve positioning

kmanijak committed:
Revert "improve positioning"

This reverts commit bfedcfbd4b744e339dbb795f7f71a09d19f842fa.

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-collection/style.scss:
```diff
@@ -2,6 +2,27 @@
+	:has(.is-product-collection-layout-carousel) {
+		position: relative;
+
+		.wc-block-next-previous-buttons {
```
I think this could also be wrapped in `:where()`, no?

Aljullu reviewed (commented):
Really cool how it's looking! :clap: 

I noticed a couple of things:

1. In RTL languages, I would expect the arrows to be in the oposite position:

https://github.com/user-attachments/assets/b6c31393-7876-47c0-a7f1-48da9d2c1267

2. The _Sale_ badge seems to have a higher z-index than the arrows, so it's possible they might get covered (it can be seen in the video above).
3. There is this _Stretch to fill_ option in the arrows block. Is it intended to have any specific effect? For me, it looks like aligned to the middle in the editor and aligned to the top in the frontend. :thinking: 

<img width="510" height="503" alt="imatge" src="https://github.com/user-attachments/assets/d0f6cea6-f53a-4ede-8698-80b83586e396" />

kmanijak committed:
Merge branch 'trunk' into wooplug-4885-product-collection-carousel-layout-frontend-logic

tyxla reviewed (commented):
Is it expected that the prev/next buttons will overlap the actual carousel by default? If yes, it could be a good opportunity to revisit the default styles there.

kmanijak committed:
Increase z-index of Next Prev Buttons

kmanijak committed:
Update the logic of inserting the arrows

kmanijak committed:
Cleanup styles

kmanijak committed:
WIP: Add some layout attributes to manipulated blocks

kmanijak committed:
Merge branch 'trunk' into wooplug-4885-product-collection-carousel-layout-frontend-logic

kmanijak committed:
Fix lint

kmanijak committed:
Merge branch 'trunk' into wooplug-4885-product-collection-carousel-layout-frontend-logic

kmanijak committed:
Adjust group layout

kmanijak committed:
Move Product Gallery specific styles to its styles

kmanijak committed:
Add proper layout values to group blocks

kmanijak committed:
Refactor the logic a bit

kmanijak committed:
Product Collection carousel: indicate scrolling by always showing a bit of next product (#60123)

* Add opinionated styles to carousel products width so it always indicates there's more to scroll

* Add changelog

* Handle the max-width if there's not enough products in carousel

* Update values to show more of the next product

kmanijak committed:
Mock SVG file

kmanijak committed:
Inner blocks in tests should be created from inner blocks template and not expect already created blocks

kmanijak committed:
Add integration tests

kmanijak committed:
Fix lint

kmanijak committed:
Update the style for mobile hiding srrows

kmanijak committed:
Merge branch 'trunk' into wooplug-4885-product-collection-carousel-layout-frontend-logic

kmanijak commented:
@Aljullu , I've updated the PR:

1. Addressed your comments and feedback 
2. Changed the arrows position: it's no longer overlapping with products. Now it's placed above the Product Template on the right, either with heading or without, depending if it's there or not. I updated the testing steps.
3. Added tests for transition from and to carousel layout

Could I ask you for re-review?

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-collection/edit/inspector-controls/use-carousel-layout-adjustments.ts:
```diff
@@ -15,79 +15,140 @@ import {
-	);
+	// 2. Create and insert the next/previous buttons block
+	const nextPrevArrowsBlock = createBlock( nextPreviousButtonsBlockName, {
+		layout: { type: 'flex' },
```
Should we add `flexWrap:"nowrap"` here? Otherwise if the text next to the arrows is long, the arrows break in two lines by default:

<img width="916" height="417" alt="imatge" src="https://github.com/user-attachments/assets/2a5f5965-061c-4a72-a8ff-a12fea09ad60" />

kmanijak replied:
Good point! Added in https://github.com/woocommerce/woocommerce/pull/60060/commits/59afcec8d78fd4926e4528ffd69e83da5e0b0571

Aljullu commented on plugins/woocommerce/client/blocks/assets/js/blocks/product-collection/frontend.ts:
```diff
@@ -35,19 +36,89 @@ function isValidLink( ref: HTMLElement | null ): ref is HTMLAnchorElement {
+		return {
+			isDisabledPrevious: true,
+			isDisabledNext: true,
+		};
```
Is this returned object used anywhere?

kmanijak replied:
Nice catch! It's not and should not be here, addressed in https://github.com/woocommerce/woocommerce/pull/60060/commits/61a3e96972548d2d7f37077b3bdba31bade49e75

Aljullu reviewed (approved):
Very nice job, it's looking awesome! :clap: 

I left a couple of minor comments inline, but pre-approving. Also, not a strong opinion, but I noticed that for the disabled buttons we use `aria-disabled` instead of `disabled`. Just want to confirm this is intentional, so we want disabled arrow buttons to be focusable?

https://github.com/user-attachments/assets/f2cd10ba-9e4e-44ce-8d02-8ce47722220a

<details>
<summary>Testing environment</summary>

* WordPress 6.8.2 local dev environment
* Plugins: WooCommerce Beta Tester
* Theme: TT5
* Store creation: June 2025
* Store size: small
</details>

kmanijak committed:
Add flex no wrap to next prev arrows

kmanijak committed:
remove unused return values

kmanijak commented:
> Also, not a strong opinion, but I noticed that for the disabled buttons we use aria-disabled instead of disabled. Just want to confirm this is intentional, so we want disabled arrow buttons to be focusable?

Yes, it's intentional as a response to [specific feedback](https://github.com/woocommerce/woocommerce/pull/58505#pullrequestreview-2894726392):

> When you focus on the gallery arrow button, pressing the arrow keys till the end or start of the gallery, the focus will lost (because they got disabled) and you won't be able to navigate the gallery using arrow keys anymore

In order to address it we keep them focusable so they can still handle arrow keys press. I hope it makes sense.

kmanijak committed:
Merge branch 'trunk' into wooplug-4885-product-collection-carousel-layout-frontend-logic

kmanijak committed:
Add proper RTL support

kmanijak commented:
@Aljullu, Code Rabbit pointed our properly the RTL wasn't fully covered (even with the separate PR: https://github.com/woocommerce/woocommerce/pull/60154). I addressed the missing piece in https://github.com/woocommerce/woocommerce/pull/60060/commits/8a2472a800c8498095b234841e8e11653d731ba0.

kmanijak committed:
Add safe guard
